Sources/SwiftCoAP/SCMessage.swift

```diff
@@ -45,63 +45,77 @@ extension SCCoAPTransportLayerDelegate {
 //MARK: - SC CoAP Transport Layer Protocol declaration
 
 public protocol SCCoAPTransportLayerProtocol: AnyObject {
-    // SCClient uses this property to assign itself as delegate
-    var transportLayerDelegate: SCCoAPTransportLayerDelegate! { get set }
-    
     // `SClient` calls one of the following methods when it wants to send CoAP data.
     //
     // Only a `sendCoAPData(_ data: Data, toEndpoint endpoint: NWEndpoint)` should be implemented.
     // A method `sendCoAPData(_ data: Data, toHost host: String, port: UInt16)` has the default implementation in protocol extension
     // just converting `host` and `port` into an `NWEndpoint` object.
-    func sendCoAPData(_ data: Data, toHost host: String, port: UInt16) throws
-    func sendCoAPData(_ data: Data, toEndpoint endpoint: NWEndpoint) throws
-
+    func sendCoAPMessage(_ message: SCMessage, toHost host: String, port: UInt16, token: UInt64?, delegate: SCCoAPTransportLayerDelegate?) throws
+    func sendCoAPMessage(_ message: SCMessage, toEndpoint endpoint: NWEndpoint, token: UInt64?, delegate: SCCoAPTransportLayerDelegate?) throws
+    
+    func getMessageId(for endpoint:NWEndpoint) -> UInt16
     // Called when the transmission is over. Clear your states (e.g. close sockets)
-    func closeTransmission()
+    func closeTransmission(for endpoint: NWEndpoint, withToken: UInt64)
```

Comment:
Let me suggest adding a killswitch method like `closeAllTransmissions()` too to terminate everything at once for the transport. We can use it in our app too for the less verbose termination code for example on going to bg.

Reply:
added